File: system_design/stock_price_threshold_notification_system.py
from collections import deque
from typing import Optional

class PriceWindow:
    """
    maintains a sliding window of (timestamp, price) over the last `window_seconds`.
    Provides O(1) amortized access to the maximum and minimum price in the window
    using monotonic deques.
    """
    def __init__(self, window_seconds: int = 600):
        self.window_seconds = window_seconds
        self.max_deque = deque()  # stores (timestamp, price) decreasing order front = maximum
        self.min_deque = deque()  # stores (timestamp, price) inreasing order front = minimum
        self.triggered = False  # flag to indicate if the threshold has been triggered

    def add_price(self, timestamp: int, price: float):
        # Remove old prices from the deques
        cutoff = timestamp - self.window_seconds
        while self.max_deque and self.max_deque[0][0] < cutoff:
            self.max_deque.popleft()
        while self.min_deque and self.min_deque[0][0] < cutoff:
            self.min_deque.popleft()
        while self.max_deque and self.max_deque[-1][1] <= price:
            self.max_deque.pop()
        self.max_deque.append((timestamp, price))
        while self.min_deque and self.min_deque[-1][1] >= price:
            self.min_deque.pop()
        self.min_deque.append((timestamp, price))

    def get_max(self) -> Optional[float]:
        return self.max_deque[0][1] if self.max_deque else None
    def get_min(self) -> Optional[float]:
        return self.min_deque[0][1] if self.min_deque else None
    def check_change(self, threshold: float = 0.10) -> bool:
        """
        Check if the price change in the window exceeds the threshold.
        Returns True if the change exceeds the threshold, False otherwise.
        """
        if not self.max_deque or not self.min_deque:
            return False
        max_price = self.max_deque[0][1]
        min_price = self.min_deque[0][1]
        if min_price == 0:
            return False  # Avoid division by zero
        change = (max_price - min_price) / min_price
        if change >= threshold and not self.triggered:
            self.triggered = True
            return True
        elif change < threshold*0.9:
            self.triggered = False  # Reset the triggered flag if the change drops below 90% of the threshold
        return False
    

File: system_design/stock_price_threshold_notification_system.cpp
#include <deque>
#include <utility>
#include <optional>
#include <cmath>

using namespace std;

class PriceWindow{
    int windowSeconds;
    deque<pair<long long, double>> maxDeque; // pair of timestamp and price
    deque<pair<long long, double>> minDeque; // pair of timestamp and price
    bool triggered = false; // flag to indicate if the threshold has been triggered

public:
    explicit PriceWindow(int windowSeconds = 600) : windowSeconds(windowSeconds) {}

    void addPrice(long long timestamp, double price) {
        // Remove old prices from the deques
        while (!maxDeque.empty() && maxDeque.front().first <= timestamp - windowSeconds) {
            maxDeque.pop_front();
        }
        while (!minDeque.empty() && minDeque.front().first <= timestamp - windowSeconds) {
            minDeque.pop_front();
        }

        // Add new price to the deques
        while (!maxDeque.empty() && maxDeque.back().second < price) {
            maxDeque.pop_back();
        }
        maxDeque.emplace_back(timestamp, price);

        while (!minDeque.empty() && minDeque.back().second > price) {
            minDeque.pop_back();
        }
        minDeque.emplace_back(timestamp, price);
    }
    optional<double> getMaxPrice() const {
        if (maxDeque.empty()) return nullopt;
        return maxDeque.front().second;
    }
    optional<double> getMinPrice() const {
        if (minDeque.empty()) return nullopt;
        return minDeque.front().second;
    }
    bool checkChange(double threshold = 0.10) {
        if (maxDeque.empty() || minDeque.empty()) return false;

        double maxPrice = maxDeque.front().second;
        double minPrice = minDeque.front().second;

        // Calculate percentage change
        double percentageChange = (maxPrice - minPrice) / minPrice;

        // Check if the change exceeds the threshold
        if (percentageChange >= threshold && !triggered) {
            triggered = true;
            return true;
        }
        else if(percentageChange < threshold && triggered) {
            triggered = false; // Reset the trigger if the change goes below the threshold
        }
        return false;
    }
};